File: MetaTracksAcoustic/AudioRecognitionLibrary/AudioRecognitionLibrary/MinHash.swift
//
// Created by metatracks on 24/04/16.
// Copyright (c) 2016 metatracks. All rights reserved.
//

import Foundation

public class MinHash {
    var HashBucketSize:Int = 100000
    var PrimeP:Int64 = 2147483659
    var A:Int = 1
    var B:Int = 0
    var permutations:[[Int]] = []
    var permutationsCount:Int = 0
    let maxNumber = 8
    
    init(){
        permutations = DefaultPermutations.GetDefaultPermutations()
    }

    public func ComputeMinHashSignatureByte(fingerprint: [Bool]) -> [UInt8] {
        var signature:[Bool] = fingerprint
        var perms:[[Int]] = permutations
        var minHash:[UInt8] = [UInt8](count: perms.count, repeatedValue: 0)
        for (var i = 0; i < perms.count; i++){
            minHash[i] = 255
            for (var j = 0; j < perms[i].count; j++){
                if(signature[perms[i][j]]){
                    minHash[i] = UInt8(j)
                    break
                }
            }

        }
        return minHash
    }

    public func GroupMinHashToLshBucketsByte(minHashes: [UInt8], numberOfHashTables: Int, numberOfMinHashesPerKey: Int) -> ([Int], [Int64]){
        var iterator:[Int] = []
        var result:[Int64] = []
        
        
        for (var i = 0; i < numberOfHashTables; i++){
            var array = [UInt8](count: maxNumber, repeatedValue: 0)
            for (var j = 0; j < numberOfMinHashesPerKey; j++){
                array[j] = minHashes[i * numberOfMinHashesPerKey + j]
            }
            
            var hashbucket = UnsafePointer<Int64>(array).memory
            hashbucket = (Int64(A) * hashbucket + Int64(B) % Int64(PrimeP)) % Int64(HashBucketSize)
            iterator.append(i)
            result.append(hashbucket)
        }
        return (iterator, result)
    }
}

File: MetaTracks/AcousticFingerprintingLibrary_0.4.5/Hashing/MinHash.cs
using System;
using System.Collections.Generic;

namespace AcousticFingerprintingLibrary_0._4._5.Hashing
{
    /// <summary>
    ///  Class for Min Hash algorithm implementation
    /// </summary>
    public class MinHash
    {
        /// <summary>
        ///   Maximum number of hash buckets in the database
        /// </summary>
        private const int HashBucketSize = 100000;
        
        // Highest primenumber
        private const long PrimeP = 2147483659;

        /// <summary>
        ///   A Constant used in computation of  hash bucket value
        /// </summary>
        private const int A = 1;

        /// <summary>
        ///   B Constant used in computation of hash bucket value
        /// </summary>
        private const int B = 0;

        /// <summary>
        ///   Permutations dictionary
        /// </summary>
        private readonly int[][] _permutations;

        /// <summary>
        ///   Number of permutation read from the database
        /// </summary>
        /// <remarks>
        ///   Default = 100
        /// </remarks>
        private readonly int _permutationsCount;

        /// <summary>
        ///   Public constructor 
        ///  perm length = 100, int[255] in each
        /// </summary>
        public MinHash()
        {
            _permutations = DefaultPermutations.GetDefaultPermutations();

            if (_permutations == null || _permutations.Length == 0)
                throw new Exception("Permutations are null or not enough to create the Min Hash signature");

            _permutationsCount = _permutations.Length;
        }

        /// <summary>
        ///   Number of random permutations
        /// </summary>
        public int PermutationsCount => _permutationsCount;

        public byte[] ComputeMinHashSignatureByte(bool[] fingerprint)
        {
            bool[] signature = fingerprint;
            int[][] perms = _permutations;
            byte[] minHash = new byte[perms.Length]; /*100*/
            for (int i = 0; i < perms.Length /*100*/; i++)
            {
                minHash[i] = 255; /*The probability of occurrence of 1 after position 255 is very insignificant*/
                for (int j = 0; j < perms[i].Length /*256*/; j++)
                {
                    if (signature[perms[i][j]])
                    {
                        minHash[i] = (byte)j; /*Looking for first occurrence of '1'*/
                        break;
                    }
                }
            }

            return minHash; /*Array of 100 elements with bit turned ON if permutation captured successfully a TRUE bit*/
        }

        public Dictionary<int, long> GroupMinHashToLshBucketsByte(byte[] minHashes, int numberOfHashTables, int numberOfMinHashesPerKey)
        {
            Dictionary<int, long> result = new Dictionary<int, long>();
            const int maxNumber = 8; /*Int64 biggest value for MinHash*/
            if (numberOfMinHashesPerKey > maxNumber)
                throw new ArgumentException("numberOfMinHashesPerKey cannot be bigger than 8");
            for (int i = 0; i < numberOfHashTables /*hash functions*/; i++)
            {
                byte[] array = new byte[maxNumber];
                for (int j = 0; j < numberOfMinHashesPerKey /*r min hash signatures*/; j++)
                {
                    array[j] = minHashes[i * numberOfMinHashesPerKey + j];
                }
                long hashbucket = BitConverter.ToInt64(array, 0); //actual value of the signature
                hashbucket = ((A * hashbucket + B) % PrimeP) % HashBucketSize;
                result.Add(i, hashbucket);
            }
            return result;
        }
    }
}